### PR TITLE
Blockquote tags

### LINKED ISSUE #718: Note, important and warning blockquote
Hello,

It’s a new github flavored markdown feature: https://github.com/orgs/community/discussions/16925 (still in beta).

Is it possible to add this?

Markdown:
```markdown
> [!NOTE]  
> Highlights information that users should take into account, even when skimming.

> [!IMPORTANT]  
> Crucial information necessary for users to succeed.

> [!WARNING]  
> Critical content demanding immediate user attention due to potential risks.
```

Rendered:
> [!NOTE]  
> Highlights information that users should take into account, even when skimming.

> [!IMPORTANT]  
> Crucial information necessary for users to succeed.

> [!WARNING]  
> Critical content demanding immediate user attention due to potential risks.

### PR DESCRIPTION
**Pull request by Martin1887**
Fixes #718.

The tags matching in `scanners.rs` and the detection of being currently in a blockquote of `firstpass.rs` could probably be improved.

More importantly, one of the tests currently fails with two consecutive blockquotes because the end of the previous one is not detected, so tags are not searched in the latest one.

### CODE REVIEW

**Comment by Martin1887** [comment]
Thanks! I will apply your reviews this week. About the options flag, I think it is not necessary for these kind of things that do not produce breaking changes (who would want to literally put in the first line of a blockquote `[!TIP]`?

File: pulldown-cmark/src/scanners.rs [hunk 1]
```diff
@@ -187,15 +187,56 @@ impl<'a> LineStart<'a> {
...
             let _ = self.scan_space(1);
-            true
+            if scan_tag {
+                let saved_ix = self.ix;
+                if self.scan_ch(b'[') && self.scan_ch(b'!') {
+                    let tag = self.bytes[self.ix..]
+                        .iter()
+                        .take_while(|&&c| c != b'\n' && c != b'\r')
+                        .copied()
+                        .collect::<Vec<u8>>();
```
[the remark this answers is not in the document]
Comment (ollpu) [reply pulldown-cmark/src/scanners.rs:205]: Or using the existing machinery
```suggestion
                    let tag_len = scan_while(&self.bytes[self.ix..], |c| c != b'\n' && c != b'\r');
                    let tag = &self.bytes[self.ix..self.ix + tag_len];
```

**Comment by ollpu** [comment]
So GFM seems to allow having empty lines (containing only whitespace) and indentation in the blockquote before the tag:

```
>        
>  [!NOTE]
>  Content
>
```

>        
>  [!NOTE]
>  Content
>

This seems like a style we also want to support (having extra empty lines or extra spaces).

It's also pretty obvious that GitHub does this in post processing. I thought maybe we should do it like that as well. This isn't great behavior though:

```
> \[!NOTE\]\
> No way to opt out?
```
> \[!NOTE\]\
> No way to opt out?

So maybe not.

---

Regarding the `Option` flag, I feel that we should keep everything strictly in line with Commonmark when no option flags are active. It might not be a breaking change for existing content in practice, but the fact that an untrusted user can now use this syntax can be unexpected for a maintainer. It also causes problems with differential fuzzing and the like.

Some of these miscellaneous GFM features could be grouped under the same GFM flag though.

**Comment by ollpu** [comment]
As for the classes emitted by the renderer, maybe they should be more specific? GitHub uses `markdown-alert markdown-alert-note` etc.

File: pulldown-cmark/src/scanners.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Martin1887) [reply pulldown-cmark/src/scanners.rs:205]: Smart, thanks!

**Comment by Martin1887** [comment]
OK, I forgot there is a flag for GFM :sweat_smile:, it must be used. I will handle the any number of spaces and tests will be added for this case too. Regarding the other implementations don't support an empty line before, I think that should not be supported.

Thanks!

**Comment by ollpu** [comment]
@notriddle Gotcha. The stricter semantics make sense then. Yet another feature where GitHub's lazy implementations has very different semantics from the rest :D 

@Martin1887 Sorry I was unclear, there's no existing flag for GFM, but one could be added if there are many miscellaneous GFM-esque features.

**Comment by ollpu** [comment]
Also I guess the extra spaces aren't supported by most other implementations either. commonmark-hs only accepts one space after, but none before. So the behavior in this PR is probably mostly ok.

**Comment by Martin1887** [comment]
I think that all issues are solved now and the pull request can be merged. I finally added an option tag for GFM misc features, currently only for blockquote tags. Thanks!

The 0.11 release is closer now!

**Comment by Martin1887** [comment]
Thanks, I have checked your code and merge it. The only thing I have changed is a comment about the while loop is wrong, left after copying my reviewed code I guess.

So this should be good to be merged now.

**Pull request merged**